javascript/ql/src/semmle/javascript/GlobalAccessPaths.qll

```diff
@@ -419,4 +419,78 @@ module AccessPath {
     or
     result = node.getALocalSource()
   }
+
+  /**
+   * A module for reasoning dominating reads and writes to access-paths.
+   */
+  module DominatingPaths {
+    /**
+     * A classification of acccess paths into reads and writes.
+     */
+    cached
+    private newtype AccessPathKind =
+      AccessPathRead() or
+      AccessPathWrite()
+
+    /**
+     * Gets the `ranking`th access-path with `root` and `path` within `bb`.
+     * And the access-path has type `type`.
+     */
+    private DataFlow::Node rankedAccessPath(
+      ReachableBasicBlock bb, Root root, string path, int ranking, AccessPathKind type
+    ) {
+      result =
+        rank[ranking](DataFlow::Node ref |
+          ref = getAccessTo(root, path, _) and
+          ref.getBasicBlock() = bb
+        |
+          ref order by any(int i | ref.asExpr().getEnclosingStmt() = bb.getNode(i))
```

Comment:
Why `getEnclosingStmt()`?

Reply:
It was a workaround to not mark the rhs of `obj.prop4 = obj.prop4;` as being dominated by the lhs.  
But the real fix for that was the `getRhs()` in your suggestion, so the `getEnclosingStmt()` is no longer needed.